manager/common/license.go

```diff
@@ -0,0 +1,207 @@
+	defer c.counterMutex.Unlock()
+	if c.licenses[vendor].Used-num < 0 {
+		log.Printf("Cannot release a negative number of licenses \n")
+		return c.licenses[vendor].Total
+	}
+	log.Printf("License released by %s \n", hash)
+	c.licenses[vendor].Used -= num
+	return c.licenses[vendor].Total
+}
+
+func (c *LicenseCounter) peekLicense(vendor string) (int, int, bool) {
+	c.counterMutex.Lock()
+	defer c.counterMutex.Unlock()
+	return c.licenses[vendor].Used, c.licenses[vendor].Total, c.licenses[vendor].Used < c.licenses[vendor].Total
+}
+
+func (c *LicenseCounter) enqueue(key string, val *Client) {
+	c.clientMutex.Lock()
+	defer c.clientMutex.Unlock()
+	c.queue[val.Vendor] = append(c.queue[val.Vendor], key)
+	c.clients[key] = val
+	log.Printf("Client %s from %s@%s waiting for %d %s %s \n",
+		key, val.User, val.IP, val.Quantity, val.Vendor, val.Feature)
+	log.Printf("%s job queue: %s \n", val.Vendor, strings.Join(c.queue[val.Vendor], ", "))
+}
+
+func (c *LicenseCounter) dequeue(key string, vendor string) {
+	c.clientMutex.Lock()
+	defer c.clientMutex.Unlock()
+	c.queue[vendor] = c.queue[vendor][1:]
+	delete(c.clients, key)
+	log.Printf("Client %s removed from the %s queue \n", key, vendor)
+}
+
+func (c *LicenseCounter) del(key string, vendor string) {
+	index := -1
+	c.clientMutex.Lock()
+	defer c.clientMutex.Unlock()
+	_, ok := c.queue[vendor]
+	if !ok {
+		log.Printf("Vendor \"%s\" does not exist in the queue \n", vendor)
+		return
+	}
+	for i := 0; i < len(c.queue[vendor]); i++ {
+		if c.queue[vendor][i] == key {
+			c.queue[vendor] = append(c.queue[vendor][:i], c.queue[vendor][i+1:]...)
+			delete(c.clients, key)
+			log.Printf("Job cancelled by client %s - removed from queue \n", key)
+			index = i
+			break
+		}
+	}
+	if index == -1 {
+		log.Printf("%s is not in queue \n", key)
+	}
+}
+
+func (c *LicenseCounter) validLicense(vendor string, feature string) bool {
+	licenses := map[string]map[string]bool{"xilinx": {"vivado": true},
+		"cadence": {"xcelium": true}}
+	_, ok := licenses[vendor][feature]
+	return ok
+}
+
+func (c *LicenseCounter) peekQueue(vendor string, hash string) bool {
+	c.clientMutex.Lock()
+	defer c.clientMutex.Unlock()
+	return c.queue[vendor][0] == hash
+}
+
+type Client struct {
+	IP       string
+	User     string
+	Start    time.Time
+	Quantity int32
+	Vendor   string
+	Feature  string
+}
+
+type License struct {
+	Total int
+	Used  int
+}
+
+var totalCadenceLic = 3
+var totalXilinxLic = 3
+var licenseCounter = LicenseCounter{
+	licenses: map[string]*License{"xilinx": &License{Used: 0, Total: totalXilinxLic}, "cadence": &License{Used: 0, Total: totalCadenceLic}},
+	clients:  map[string]*Client{},
+	queue:    map[string][]string{"xilinx": make([]string, 0), "cadence": make([]string, 0)}}
+
+type Server struct {
+	rng *rand.Rand
+}
+
+func (s *Server) Polling(stream rpc_license.License_PollingServer) error {
+	hash := ""
+	vendor := ""
+	var status error
+	s.rng = rand.New(rand.NewSource(rand.Int63()))
+	for {
```

Comment:
1. If the connection with the client terminates prematurely before a license is allocated, remove the client from the queue but do not decrement the license counter.
2. If the connection with the client terminates prematurely AFTER the license has been allocated or if the client finishes its job gracefully (implying that the client has also been removed from the queue already), only decrement the license counter.

Because disconnection or failure on the client side can only happen in these 2 states, I cannot call `defer` because I need to handle these two different scenarios differently.